File: aoc-2015/src/test/java/com/capital7software/aoc2015/days/Day21Test.java
package com.capital7software.aoc2015.days;

import com.capital7software.aoc.aoc2015.days.Day21;
import com.capital7software.aoc.lib.AdventOfCodeTestBase;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.logging.Logger;

import static org.junit.jupiter.api.Assertions.assertEquals;

class Day21Test extends AdventOfCodeTestBase {
    private static final Logger LOGGER = Logger.getLogger(Day21Test.class.getName());

    @BeforeEach
    void setUp() {
        var instance = new Day21();
        setupFromFile(instance.getDefaultInputFilename());
    }

    @Test
    void leastAmountOfGoldAndStillWin() {
        var instance = new Day21();
        int expected = 65;
        var actual = instance.leastAmountOfGoldAndStillWin(lines);
        assertEquals(expected, actual, "The least amount of gold " +
                "is not what was expected: " + expected);
    }

    @Test
    void mostAmountOfGoldAndStillLose() {
        var instance = new Day21();
        int expected = 188;
        var actual = instance.mostAmountOfGoldAndStillLose(lines);
        assertEquals(expected, actual, "The most amount of gold " +
                "is not what was expected: " + expected);
    }

    @Override
    protected Logger getLogger() {
        return LOGGER;
    }
}

File: aoc-2024/src/test/kotlin/com/capital7software/aoc/aoc2024aoc/days/Day21Test.kt
package com.capital7software.aoc.aoc2024aoc.days

import com.capital7software.aoc.lib.AdventOfCodeTestBase
import com.capital7software.aoc.lib.grid.KeypadConundrum
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class Day21Test : AdventOfCodeTestBase() {
  companion object {
    val log: Logger = LoggerFactory.getLogger(Day21Test::class.java)
  }

  override fun getLogger(): Logger {
    return log
  }

  @BeforeEach
  fun setup() {
    val instance = Day21()
    setupFromFile(instance.defaultInputFilename)
  }

  @Test
  fun testShortestHumanInputLength() {
    val conundrum = KeypadConundrum.parse(lines)

    val expected = mapOf(
        "029A" to 68L,
        "980A" to 60L,
        "179A" to 68L,
        "456A" to 64L,
        "379A" to 64L
    )

    expected.forEach { (code, length) ->
      Assertions.assertEquals(
          length,
          conundrum.shortestHumanInputLength(code, directionalRobots = 2),
          "The shortest human input length for $code is not the expected value $length."
      )
    }
  }

  @Test
  fun testPartOneExample() {
    val actual = Day21().sumOfComplexities(lines, directionalRobots = 2)

    Assertions.assertEquals(126384L, actual)
  }

  @Test
  fun testPartOneConvenienceMethodExample() {
    val actual = KeypadConundrum.part1(lines)

    Assertions.assertEquals(126384L, actual)
  }

  @Test
  fun testPartTwoExample() {
    val actual = Day21().sumOfComplexities(lines, directionalRobots = 25)

    Assertions.assertEquals(154115708116294L, actual)
  }

  @Test
  fun testPartTwoConvenienceMethodExample() {
    val actual = KeypadConundrum.part2(lines)

    Assertions.assertEquals(154115708116294L, actual)
  }
}
